File: javapatterns/src/main/java/org/iut/nantes/factories/cours/FactoryCours.java
package org.iut.nantes.factories.cours;

abstract class FactoryCours
{
  public abstract Cours creerCours();
  public abstract Prof creerProf(String nom );
  public abstract Salle creerSalle();
}

class FactoryCoursInfo extends FactoryCours
{
  public Cours creerCours() { return new CoursInfo(); }
  public Prof creerProf(String nom ) { return new ProfInfo( nom ); }
  public Salle creerSalle() { return new SalleInfo(); }
}

class FactoryCoursMaths extends FactoryCours
{
  public Cours creerCours() { return new CoursMaths(); }
  public Prof creerProf(String nom ) { return new ProfMaths( nom ); }
  public Salle creerSalle() { return new SalleMaths(); }
}

class FactoryCoursChimie extends FactoryCours
{
  public Cours creerCours() { return new CoursChimie(); }
  public Prof creerProf(String nom ) { return new ProfChimie( nom ); }
  public Salle creerSalle() { return new SalleChimie(); }
}

File: kotlinpatterns/src/main/kotlin/org/iut/nantes/factories/cours/FactoryCours.kt
package org.iut.nantes.factories.cours

abstract class FactoryCours {
    abstract fun creerCours(): Cours
    abstract fun creerProf(nom: String): Prof
    abstract fun creerSalle(): Salle
}

class FactoryCoursInfo : FactoryCours() {
    override fun creerCours(): Cours {
        return CoursInfo()
    }

    override fun creerProf(nom: String): Prof {
        return ProfInfo(nom)
    }

    override fun creerSalle(): Salle {
        return SalleInfo()
    }
}

class FactoryCoursMaths : FactoryCours() {
    override fun creerCours(): Cours {
        return CoursMaths()
    }

    override fun creerProf(nom: String): Prof {
        return ProfMaths(nom)
    }

    override fun creerSalle(): Salle {
        return SalleMaths()
    }
}

class FactoryCoursChimie : FactoryCours() {
    override fun creerCours(): Cours {
        return CoursChimie()
    }

    override fun creerProf(nom: String): Prof {
        return ProfChimie(nom)
    }

    override fun creerSalle(): Salle {
        return SalleChimie()
    }
}